sw/device/lib/dif/dif_keymgr.c

```diff
@@ -5,3 +5,710 @@
+      return false;
+  }
+
+  return true;
+}
+
+/**
+ * Parameters for starting an operation.
+ *
+ * Values of the members must be the actual values that will be written to the
+ * CONTROL register.
+ */
+typedef struct start_operation_params {
+  /**
+   * Destination for this operation.
+   */
+  uint32_t dest;
+  /**
+   * Operation to start.
+   */
+  uint32_t op;
+} start_operation_params_t;
+
+/**
+ * Starts an operation.
+ */
+static void start_operation(const dif_keymgr_t *keymgr,
+                            start_operation_params_t params) {
+  uint32_t reg_control =
+      bitfield_field32_write(0, KEYMGR_CONTROL_DEST_SEL_FIELD, params.dest);
+  reg_control = bitfield_field32_write(
+      reg_control, KEYMGR_CONTROL_OPERATION_FIELD, params.op);
+  reg_control =
+      bitfield_bit32_write(reg_control, KEYMGR_CONTROL_START_BIT, true);
+  mmio_region_write32(keymgr->params.base_addr, KEYMGR_CONTROL_REG_OFFSET,
+                      reg_control);
+}
+
+/**
+ * Returns the bit index for a given IRQ.
+ */
+DIF_WARN_UNUSED_RESULT
+static bool get_irq_bit_index(dif_keymgr_irq_t irq,
+                              bitfield_bit32_index_t *bit_index) {
+  switch (irq) {
+    case kDifKeymgrIrqDone:
+      *bit_index = KEYMGR_INTR_COMMON_OP_DONE_BIT;
+      break;
+    default:
+      return false;
+  }
+
+  return true;
+}
+
+/**
+ * Checks if a value is a valid `dif_keymgr_toggle_t` and converts it to `bool`.
+ */
+DIF_WARN_UNUSED_RESULT
+static bool toggle_to_bool(dif_keymgr_toggle_t val, bool *val_bool) {
+  switch (val) {
+    case kDifKeymgrToggleEnabled:
+      *val_bool = true;
+      break;
+    case kDifKeymgrToggleDisabled:
+      *val_bool = false;
+      break;
+    default:
+      return false;
+  }
+  return true;
+}
+
+/**
+ * Converts a `bool` to `dif_keymgr_toggle_t`.
+ */
+static dif_keymgr_toggle_t bool_to_toggle(bool val) {
+  return val ? kDifKeymgrToggleEnabled : kDifKeymgrToggleDisabled;
+}
+
+dif_keymgr_result_t dif_keymgr_init(dif_keymgr_params_t params,
+                                    dif_keymgr_t *keymgr) {
+  if (keymgr == NULL) {
+    return kDifKeymgrBadArg;
+  }
+
+  *keymgr = (dif_keymgr_t){.params = params};
+
+  return kDifKeymgrOk;
+}
+
+dif_keymgr_result_t dif_keymgr_configure(const dif_keymgr_t *keymgr,
+                                         const dif_keymgr_config_t *config) {
+  if (keymgr == NULL || config == NULL) {
+    return kDifKeymgrBadArg;
+  }
+
+  uint32_t reg_val = bitfield_field32_write(0, KEYMGR_RESEED_INTERVAL_VAL_FIELD,
+                                            config->entropy_reseed_interval);
+  mmio_region_write32(keymgr->params.base_addr,
+                      KEYMGR_RESEED_INTERVAL_REG_OFFSET, reg_val);
+
+  return kDifKeymgrOk;
+}
+
+dif_keymgr_lockable_result_t dif_keymgr_advance_state(
+    const dif_keymgr_t *keymgr, const dif_keymgr_state_params_t *params) {
+  if (keymgr == NULL) {
+    return kDifKeymgrLockableBadArg;
+  }
+
+  if (!is_idle(keymgr)) {
```

Comment:
> Is there a distinction between idle and locked? I am just asking because of the function name and then the return value. 

In this case, not really. Ideally, the caller must make sure that the keymgr is idle before calling this function and config registers should be locked whenever the keymgr is busy. `is_idle` checks both (keymgr is idle and config registers are unlocked) to be safe but I didn't define a separate return code for `busy` to keep things simple.

> Is it possible to run into a situation where the keymgr is WIP at the time this function is called?

Yes, if the caller starts another operation before calling this function and doesn't check the status of the keymgr. The status register is rw1c. This check makes it harder to drop status codes between operations.

Please let me know if this makes sense.

Reply:
I agree, `is_ready` sounds better.